src/swap/defi/defiUtils.ts

```diff
@@ -34,29 +19,36 @@ export interface InOutTokenAddresses {
  * contract addresses, if different.
  */
 export const getInOutTokenAddresses = (
-  currencyInfo: EdgeCurrencyInfo,
-  fromCurrencyCode: string,
-  toCurrencyCode: string
+  currencyConfig: EdgeCurrencyConfig,
+  wrappedMainnetAddress: string,
+  fromTokenId?: string,
+  toTokenId?: string
 ): InOutTokenAddresses => {
-  const { currencyCode: nativeCurrencyCode, metaTokens } = currencyInfo
-  const wrappedCurrencyCode = `W${nativeCurrencyCode}`
-  const isFromNativeCurrency = fromCurrencyCode === nativeCurrencyCode
-  const isToNativeCurrency = toCurrencyCode === nativeCurrencyCode
-  const isFromWrappedCurrency = fromCurrencyCode === wrappedCurrencyCode
-  const isToWrappedCurrency = toCurrencyCode === wrappedCurrencyCode
+  const { allTokens } = currencyConfig
+
+  const isFromNativeCurrency = fromTokenId == null
+  const isToNativeCurrency = toTokenId == null
+
+  const fromToken: EdgeToken | undefined = isFromNativeCurrency
+    ? undefined
+    : allTokens[fromTokenId]
+  const toToken: EdgeToken | undefined = isToNativeCurrency
+    ? undefined
+    : allTokens[toTokenId]
+
+  const fromTokenAddress: string =
+    fromToken?.networkLocation?.contractAddress ?? ''
+  const toTokenAddress: string = toToken?.networkLocation?.contractAddress ?? ''
```

Comment:
Since the `networkLocation` parameter is untyped, we should really use a cleaner:

```js
const asContractLocation = asObject({
  contractAddress: asString
})

const fromTokenAddress = asMaybeContractLocation(toToken?.networkLocation)?.contractAddress ?? ''
```

Other chains have a different structure, so this make the assumption explicit. Ripple tokens are totally weird, for example, and will probably need a different cleaner.